File: ios/liquid_toasts/Sources/liquid_toasts/WireDecoding.swift
import CoreGraphics
import Foundation

/// Decoding helpers for method-channel payloads. `StandardMessageCodec`
/// bridges Dart ints/doubles/bools as `NSNumber`, so plain `as?` casts on the
/// primitive types are unreliable — always go through these.
extension Dictionary where Key == String, Value == Any {
  func int(_ key: String) -> Int? {
    if let n = self[key] as? NSNumber { return n.intValue }
    return self[key] as? Int
  }

  func double(_ key: String) -> Double? {
    if let n = self[key] as? NSNumber { return n.doubleValue }
    return self[key] as? Double
  }

  func cgFloat(_ key: String) -> CGFloat? {
    double(key).map { CGFloat($0) }
  }

  func bool(_ key: String, default defaultValue: Bool) -> Bool {
    if let n = self[key] as? NSNumber { return n.boolValue }
    return self[key] as? Bool ?? defaultValue
  }

  func string(_ key: String) -> String? { self[key] as? String }

  /// Decodes a wire enum (Dart `.name` string) with a fallback.
  func enumValue<E: RawRepresentable>(_ key: String, default defaultValue: E) -> E
  where E.RawValue == String {
    (self[key] as? String).flatMap(E.init(rawValue:)) ?? defaultValue
  }

  /// Decodes an optional wire enum (nil when absent or unknown).
  func enumValue<E: RawRepresentable>(_ key: String) -> E?
  where E.RawValue == String {
    (self[key] as? String).flatMap(E.init(rawValue:))
  }
}

File: android/src/main/kotlin/com/rehmatsg/liquid_toasts/WireDecoding.kt
package com.rehmatsg.liquid_toasts

/**
 * Decoding helpers for method-channel payloads. Flutter's [StandardMessageCodec]
 * bridges Dart ints as either `Int` or `Long` and doubles as `Double`, so plain
 * casts on the primitive types are unreliable — always go through these. This is
 * the Android analog of `WireDecoding.swift`'s NSNumber-aware accessors.
 */
internal fun Map<String, Any?>.optString(key: String): String? = this[key] as? String

/** Reads an int, accepting the `Int` OR `Long` the codec may deliver. */
internal fun Map<String, Any?>.optInt(key: String): Int? = when (val v = this[key]) {
    is Int -> v
    is Long -> v.toInt()
    is Number -> v.toInt()
    else -> null
}

/** Reads a double, accepting any numeric representation (`Int`/`Long`/`Double`/`Float`). */
internal fun Map<String, Any?>.optDouble(key: String): Double? = when (val v = this[key]) {
    is Double -> v
    is Float -> v.toDouble()
    is Int -> v.toDouble()
    is Long -> v.toDouble()
    is Number -> v.toDouble()
    else -> null
}

internal fun Map<String, Any?>.optBool(key: String, default: Boolean): Boolean =
    when (val v = this[key]) {
        is Boolean -> v
        else -> default
    }

@Suppress("UNCHECKED_CAST")
internal fun Map<String, Any?>.optMap(key: String): Map<String, Any?>? =
    this[key] as? Map<String, Any?>

/** Image bytes arrive as a `ByteArray` (from a Dart `Uint8List`). */
internal fun Map<String, Any?>.byteArray(key: String): ByteArray? = this[key] as? ByteArray

/**
 * Decodes a wire enum by its Dart `.name` string, falling back to [default] when
 * the key is absent or holds an unknown value. [values] is the enum's entries
 * (pass `EnumType.entries`); each entry's wire name comes from [wireName].
 */
internal fun <E> Map<String, Any?>.enumByWireName(
    key: String,
    values: List<E>,
    wireName: (E) -> String,
    default: E,
): E {
    val raw = this[key] as? String ?: return default
    return values.firstOrNull { wireName(it) == raw } ?: default
}
